### PR TITLE
imagemagick: reintroduce package, cleanup and upgrade

### PR DESCRIPTION
**Pull request by val-kulkov**
Maintainer: me
Compile tested: ipq806x, TP-Link Archer C2600, OpenWrt SNAPSHOT r9257-5a8d03c; bcm53xx, Buffalo WXR-1900DHP, OpenWrt SNAPSHOT r9257-5a8d03c; x86_64, VirtualBox VM, OpenWrt SNAPSHOT r9257-5a8d03c
Run tested: ipq806x, TP-Link Archer C2600, OpenWrt SNAPSHOT r9257-5a8d03c; bcm53xx, Buffalo WXR-1900DHP, OpenWrt SNAPSHOT r9257-5a8d03c; x86_64, VirtualBox VM, OpenWrt SNAPSHOT r9257-5a8d03c

Description:

ImageMagick is a free and open-source software suite for displaying, converting, and editing raster image and vector image files. This package provides the prominent 'convert' utility.

ImageMagick's installation footprint is:

- x86_64: 6.7 MB,
- ipq806x (ARM): 6.0 MB, 
- bcm53xx: 6.2 MB.

The shared libraries occupy 2.4 MB on bcm53xx, 2.3 MB on ipq806x, and 2.5 MB on x86_64. The 114 ImageMagick's modules occupy from 3.4 MB to 4.0 MB depending on the target. It may be possible to reduce the installation footprint by introducing build parameters to control the  selection of modules. In view of the large number of modules and the  possibility of breakage due to module interdependencies or other reasons, such attempt is not made at this time.

ImageMagick is therefore best suited for extroot-enabled or x86_64 OpenWrt systems.

In many cases, GraphicsMagick may be used as a substitute for ImageMagick. GraphicsMagick provides similar functionality, it is faster and it has a smaller installation footprint. It is therefore better suited for non-extroot OpenWrt systems. However, in tests to reduce resolution (and size) of a high-resolution JPEG image GraphicsMagick required about 25% more RAM than ImageMagick (no HDRI, quantum depth of 8) during its execution.

Signed-off-by: Val Kulkov <val.kulkov@gmail.com>

### CODE REVIEW

File: multimedia/imagemagick/Makefile [hunk 1]
```diff
@@ -0,0 +1,142 @@
+#
+# Copyright (C) 2010-2019 OpenWrt.org
+#
```
Comment (thess) [inline multimedia/imagemagick/Makefile]: Remove copyright assignment to OpenWrt.

File: multimedia/imagemagick/Makefile [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	http://mirror.checkdomain.de/imagemagick/releases \
+	ftp://ftp.fifi.org/pub/ImageMagick
+PKG_HASH:=cdf3625e72ed6e58511fdf8e32070b318793ef28e3a6c96254c64d2f6fbcff8a
+
+PKG_MAINTAINER:=Val Kulkov <val.kulkov@gmail.com>
+
+PKG_BUILD_PARALLEL:=1
+PKG_USE_MIPS16:=0
+DISABLE_NLS:=
+PKG_BUILD_DEPENDS:=zlib freetype libpng libjpeg tiff
```
Comment (thess) [inline multimedia/imagemagick/Makefile]: This is not necessary since they are already direct dependents.

File: multimedia/imagemagick/Makefile (hunk 1 above)
Comment (val-kulkov) [reply multimedia/imagemagick/Makefile]: Done. Is this now the recommended practice to avoid the OpenWrt copyright notice?

**Comment by diizzyy** [comment]
Any reason why not use graphicsmagick instead which usually is _much_ faster and also possibly smaller?

**Comment by val-kulkov** [comment]
@diizzyy : no reason other than my previous experience working with imagemagick. One day when I have some spare time I might consider porting graphicsmagick to OpenWrt. This may not happen any time soon, however.

**Comment by diizzyy** [comment]
@val-kulkov 
In that regard I'd think it would be more useful to import grapicsmagick and possibly import https://packages.debian.org/sid/graphicsmagick-imagemagick-compat if it's needed.

**Comment by val-kulkov** [comment]
@diizzyy : well, it turns out that porting GraphicsMagick is easy once you have a working Makefile for ImageMagick. Only a few modifications were required: see PR #8040.

GraphicsMagick is indeed more lightweight and it appears to be a little faster than ImageMagick. Unfortunately, it does not do everything ImageMagick does. It cannot resize image to a specified file size: "-define jpeg:extent=<X>kb" functionality is apparently missing in GM.

**Comment by DazzyWalkman** [comment]
> In that regard I'd think it would be more useful to import grapicsmagick and possibly import https://packages.debian.org/sid/graphicsmagick-imagemagick-compat if it's needed.

There is  a build flag --enable-magick-compat for GM doing almost the same thing. 
It's for the user to consider if it's needed to co-exist with imagemagick, Simply make it enabled by default, or better still, controlled by a build config option. 

But IMHO I will not bother with this  for the existing set of IM scripts' sake. 
Both can be easily ported/reintroduced to OpenWRT,  and are not much different for casual use,
But for IM scripts, some rewrites or wrapper for GM may be needed.  

GM and IM have been developed in their own ways since 2002, and have no intention to be "compatible" with each other. They do differ. Some scripts requiring IM 6 or later will likely break on GM. GM can only be considered as IM 5.5.2 in 'IM script compatiable mode'. This is a gap that can not be covered with some symlink shortcut workaround.

Users should already know both are more or less memory hungry, even if GM claims to be faster and lighter.
 
BTW. The default (unspecified) quantum depth of IM 7 is 16, while GM is 8, that makes huge difference in memory usage. I think 8 is better suit for typical OpenWRT hardware, if your scripts fail to work on GM, and then decide to give IM a shot instead of rewrite.

**Comment by val-kulkov** [comment]
Setting IM's quantum depth to 8 and disabling HDRI made a dramatic difference on the amount of RAM consumed during execution. In fact, in my tests ImageMagick now uses 25% less memory than GraphicsMagick:

**ImageMagick (quantum depth 8, no HDRI), x86_64:**
```
time -v convert img-in.jpg -resize "1200x>" -resize "x675>" -define jpeg:extent=120kb img-out.jpg
	User time (seconds): 0.71
	System time (seconds): 0.09
	Percent of CPU this job got: 63%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0m 1.26s
	Maximum resident set size (kbytes): 93248
```

**GraphicsMagick, x86_64:**
```
time -v gm convert img-in.jpg -resize "1200x675>" -quality 66 img-out.jpg
	User time (seconds): 0.37
	System time (seconds): 0.02
	Percent of CPU this job got: 99%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0m 0.39s
	Maximum resident set size (kbytes): 115248
```

**ImageMagick (quantum depth 8, no HDRI), ipq806x (ARM):**
```
time -v convert img-in.jpg -resize "1200x>" -resize "x675>" -define jpeg:extent=120kb img-out.jpg
	User time (seconds): 5.05
	System time (seconds): 0.56
	Percent of CPU this job got: 94%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0m 5.93s
	Maximum resident set size (kbytes): 91888
```

**GraphicsMagick, ipq806x (ARM):**
```
time -v gm convert img-in.jpg -resize "1200x675>" -quality 66 img-out.jpg
	User time (seconds): 1.81
	System time (seconds): 0.48
	Percent of CPU this job got: 88%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0m 2.59s
	Maximum resident set size (kbytes): 114032
```

**ImageMagick (quantum depth 8, no HDRI), bcm53xx (ARM):**
```
time -v convert img-in.jpg -resize "1200x>" -resize "x675>" -define jpeg:extent=120kb img-out.jpg
	User time (seconds): 25.79
	System time (seconds): 0.09
	Percent of CPU this job got: 99%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0m 25.91s
	Maximum resident set size (kbytes): 92592
```

**GraphicsMagick, bcm53xx (ARM):**
```
time -v gm convert img-in.jpg -resize "1200x675>" -quality 66 img-out.jpg
	User time (seconds): 17.50
	System time (seconds): 0.03
	Percent of CPU this job got: 99%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0m 17.54s
	Maximum resident set size (kbytes): 114432
```

Bottom line, GM is roughly 2x faster than IM but it consistently consumes more RAM than IM during its execution. Both packages are rather heavy and better suited for extroot-enabled devices or x86_64 OpenWrt devices rather than embedded devices. It is especially true for devices lacking FPU: on bcm53xx target which has no FPU, both GM and IM are terribly slow.

For extroot-enabled and x86_64 OpenWrt devices, the choice between IM and GM will likely be a personal one. If the primary consideration is speed, GM is the winner. If the primary consideration is the compatibility with existing scripts, then in many cases IM is going to be a likely winner. On extroot systems where RAM is of primary concern, IM is probably going to be a winner.

**Comment by mhei** [comment]
I'd like to see this getting merged, but I would like to have some nitpicks addressed before the merge. This is why I created an additional commit on top of yours. Please have a look and comment. I only compile-tested for mxs platform at the moment...

**Comment by neheb** [comment]
@val-kulkov what's the status of this?

**Comment by val-kulkov** [comment]
@mhei : thanks for the suggestions. I applied them all. I am a little doubtful about removing PKG_BUILD_DEPENDS: see [my earlier comment](https://github.com/openwrt/packages/pull/8035#discussion_r251107846). However, with parallel build removed, it probably should be fine. We'll see.

@neheb : all ready now. I rebased to head and bumped imagemagick to the latest version.

File: multimedia/imagemagick/Makefile [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+PKG_INSTALL:=1
+
+include $(INCLUDE_DIR)/package.mk
+
+define Package/imagemagick/Default
+  SECTION:=multimedia
+  CATEGORY:=Multimedia
+  TITLE:=Image manipulation tools
+  URL:=https://www.imagemagick.org/
+  DEPENDS:=imagemagick
```
Comment (neheb) [inline multimedia/imagemagick/Makefile]: Any reason this is done like this? Notice that later, you redefine DEPENDS.

File: multimedia/imagemagick/Makefile [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+PKG_VERSION:=7.0.8
+PKG_REVISION:=49
+PKG_RELEASE:=1
+
+PKG_BUILD_DIR:=$(BUILD_DIR)/ImageMagick-$(PKG_VERSION)-$(PKG_REVISION)
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION)-$(PKG_REVISION).tar.gz
+PKG_SOURCE_URL:=http://github.com/ImageMagick/ImageMagick/archive/$(PKG_VERSION)-$(PKG_REVISION)
+PKG_HASH:=53f7963bbe81520e799e9e178a13757890ed43bc9faf2e86fae1cf58aea28575
+
+PKG_LICENSE:=Apache-2.0
```
Comment (neheb) [inline multimedia/imagemagick/Makefile]: I'd put PKG_MAINTAINER above here.

File: multimedia/imagemagick/Makefile [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+#
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=imagemagick
+PKG_VERSION:=7.0.8
+PKG_REVISION:=49
+PKG_RELEASE:=1
+
+PKG_BUILD_DIR:=$(BUILD_DIR)/ImageMagick-$(PKG_VERSION)-$(PKG_REVISION)
```
Comment (neheb) [inline multimedia/imagemagick/Makefile]: I'd put PKG_BUILD_DIR below PKG_HASH

File: multimedia/imagemagick/Makefile (hunk 3 above)
Comment (val-kulkov) [reply multimedia/imagemagick/Makefile]: That was a leftover from some earlier work. Removed.

File: multimedia/imagemagick/Makefile (hunk 4 above)
Comment (val-kulkov) [reply multimedia/imagemagick/Makefile]: Fine, moved PKG_MAINTAINTER above.

File: multimedia/imagemagick/Makefile (hunk 5 above)
Comment (val-kulkov) [reply multimedia/imagemagick/Makefile]: Done.

**Comment by val-kulkov** [comment]
@neheb : you may be unable to easily see the differences because of the force push. In short, I applied all your suggestions.

File: multimedia/imagemagick/Makefile [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=imagemagick
+PKG_VERSION:=7.0.8
+PKG_REVISION:=49
+PKG_RELEASE:=1
+PKG_MAINTAINER:=Val Kulkov <val.kulkov@gmail.com>
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION)-$(PKG_REVISION).tar.gz
+PKG_SOURCE_URL:=http://github.com/ImageMagick/ImageMagick/archive/$(PKG_VERSION)-$(PKG_REVISION)
```
Comment (diizzyy) [inline multimedia/imagemagick/Makefile]: Any reason why official release tarball isn't used?
https://imagemagick.org/script/mirror.php --> http://ftp.acc.umu.se/mirror/imagemagick.org/ftp/ImageMagick-7.0.8-49.tar.xz etc

Comment (val-kulkov) [reply multimedia/imagemagick/Makefile]: Yes. As soon as a new incremental version is released, which happens about every couple of weeks with ImageMagick, the official mirrors tend to "lose" the earlier versions. When that happens, OpenWrt's robots will get 404 Not Found.

For this reason, we should stick with Github as the source repo for ImageMagick versions.
